server/datastore/mysql/apple_mdm.go

```diff
@@ -922,7 +922,8 @@ func unionSelectDevices(devices []godep.Device) (stmt string, args []interface{}
 
 func (ds *Datastore) GetNanoMDMEnrollment(ctx context.Context, id string) (*fleet.NanoEnrollment, error) {
 	var nanoEnroll fleet.NanoEnrollment
-	err := sqlx.GetContext(ctx, ds.reader, &nanoEnroll, `SELECT id, device_id, type, enabled, token_update_tally
+	// use writer as it is used just after creation in some cases
+	err := sqlx.GetContext(ctx, ds.writer, &nanoEnroll, `SELECT id, device_id, type, enabled, token_update_tally
```

Comment:
Looks good.

The issue is that we are updating the same database/tables but in two different packages, right? (which leads to this decoupling)

I wonder if we should have two methods `GetNanoMDMEnrollmentFromReader` and `GetNanoMDMEnrollmentFromWriter` even if it looks really bad just so that it's less error prone and more explicit that we need some refactoring moving forward.

Reply:
AIUI (if the fix is confirmed), it's not a question of (Go) packages, it's the sequence and speed at which two calls are made to the Fleet server that end up with `GetNanoMDMEnrollment` (or `GetHostMDMCheckinInfo`) not being able to find a row that was just created, because of the replication lag (it has been written in the primary mysql instance, but not yet replicated to the replica).

Reply:
Sorry, did not meant to refer to the bug itself but more of a "decoupling" or "error prone-ness" as an issue moving forward (where else are we missing some of these `ds.reader` -> `ds.writer`...).

Reply:
That being said I see that those methods do get called from a few places, some of which would probably not have the race issue. Maybe we can have it being reader-based when it's not in the `MDMAppleCheckinAndCommandService` flow, though it could be in a future PR so that the fix can be confirmed first.

Reply:
> where else are we missing some of these ds.reader -> ds.writer...

Yeah, that's the thing - it's technically everywhere, it just needs the replication lag to be big enough. But of course that would miss the point of taking advantage of a read replica, so we assume a reasonably small replication lag. 

Places that we "know for sure" (i.e. assume we need to read from the writer) is when we call a read method soon after a write one in the same request. Places less obvious are ones like this issue, where it's distinct requests coming in in quick succession and the replication lag is too long. Those are pretty hard to identify in general, and I'm not sure how we could automate finding those out - if we artificially slow the replication and deliberately call quick related write-read endpoints, we would trigger the issue easily, but that doesn't mean it is one in practice.

Reply:
Agreed on all! and just to over-communicate, we want to move the delivery of packages to a cron, so I'm imagining that we set a flag in the database on `Authenticate`, and the cron picks all the hosts flagged and sends the commands, which doesn't require to use `db.writer` to read results anymore.